### ISSUE TITLE
Weird runtime failures when using a channel and chained continuations

### ISSUE THREAD

**Issue by krivenko** [STEllAR-GROUP/hpx#2916]
I've been trying to implement a simple load balancing scheme, in which I dispatch actions to free localities/threads. A list of free resources is organized as a FIFO container and implemented using `hpx::lcos::channel`.

```c++
#include <hpx/hpx_init.hpp>
#include <hpx/include/lcos.hpp>
#include <hpx/include/iostreams.hpp>
#include <boost/program_options.hpp>

typedef hpx::naming::id_type locality_id_t;
HPX_REGISTER_CHANNEL(locality_id_t);

int hpx_main(boost::program_options::variables_map & p) {

    // List of currently available resources
    hpx::lcos::channel<locality_id_t> free_resources(hpx::find_here());

    int os_thread_count = hpx::get_os_thread_count();

    // At the beginning all threads on all localities are free
    for (locality_id_t id : hpx::find_all_localities()) {
        for (int i = 0; i < os_thread_count; ++i) {
            free_resources.set(id);
        }
    }

    for(int i = 0; i < 1000; ++i) {

        // Ask for resources
        hpx::shared_future<locality_id_t> target = free_resources.get();

        // Do some work, once we have acquired resources
        hpx::shared_future<int> result = target.then([](hpx::shared_future<locality_id_t> t) -> hpx::shared_future<int> {
            hpx::cout << "launching on target " << t.get() << hpx::endl;
            return hpx::make_ready_future(0);
        });

        // Free resources
        result.then([&free_resources, target](hpx::shared_future<int>){
            hpx::cout << "freeing target " << target.get() << hpx::endl;
            free_resources.set(target.get());
        });

        result.get();
    }

    free_resources.close(true);

    return hpx::finalize();
}

int main(int argc, char* argv[]) {
    boost::program_options::options_description op("HPX continuations bug");
    return hpx::init(op, argc, argv);
}
```

This program occasionally crashes with a variety of error messages.

```
{what}: assertion 'nullptr != self_.get()' failed: HPX(assertion_failure)
```
```
Stack overflow in coroutine at address 000000000000000000.

Configure the hpx runtime to allocate a larger coroutine stack size.
 Use the hpx.stacks.small_size, hpx.stacks.medium_size,
 hpx.stacks.large_size, or hpx.stacks.huge_size configuration
flags to configure coroutine stack sizes.

terminate called without an active exception
```
```
continuations_bug: /home/igor/hpx/installed/include/hpx/lcos/detail/async_implementations.hpp:72: hpx::detail::handle_managed_target<Result>::~handle_managed_target() [with Result = void]: Assertion `id_' failed.
```

My compiler is GCC 5.4.0, both HPX and the program above are built in the `Debug` mode.

P.S. Running the program on multiple localities via `mpirun` makes the crashes somewhat more probable.

**Comment by hkaiser** [comment]
@krivenko I'm currently travelling but will look into this problem asap.

**Comment by hkaiser** [comment]
@krivenko: I'm not able to reproduce the problem on my platform. I will add your test case to our buildbot testing infrastructure, let's see if the problem shows up there.

**Comment by kleinhenz** [comment]
I've been able to reproduce this in `Release` mode on cori with gcc 6.3.0 on the current hpx master.
My error message was
```
{stack-trace}: 7 frames:
0x2b0cca48e122  : ??? + 0x2b0cca48e122 in /global/homes/j/jkleinh/local/lib/libhpx.so.1
0x2b0cca4894a8  : hpx::termination_handler(int) + 0xd8 in /global/homes/j/jkleinh/local/lib/libhpx.so.1
0x2b0ccc75db10  : ??? + 0x2b0ccc75db10 in /lib64/libpthread.so.0
0x417d65        : ??? + 0x417d65 in ./build/src/issue
0x2b0cca87654b  : hpx::threads::coroutines::detail::coroutine_impl::operator()() + 0xfb in /global/homes/j/jkleinh/local/lib/libhpx.so.1
0x2b0cca87bf19  : ??? + 0x2b0cca87bf19 in /global/homes/j/jkleinh/local/lib/libhpx.so.1
{what}: Segmentation fault
```

The error is also intermittent for me.

**Comment by hkaiser** [comment]
@kleinhenz If you happen to see it in a Debug run, please post the backtrace here as well.

**Comment by sithhell** [comment]
I was able to reproduce the error locally as well with a undefined behavior sanitizer build. The error is the following:
```
/opt/apps/x86_64/clang/3.8.0/bin/../include/c++/v1/memory:4292:9: runtime error: member call on misaligned address 0x00303d64695f for type 'std::__1::__shared_weak_count', which requires 8 byte alignment
0x00303d64695f: note: pointer points here
<memory cannot be printed>
    #0 0x7f2c0b27ec89 in std::__1::shared_ptr<hpx::components::detail::fixed_wrapper_heap<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, hpx::lcos::local::spinlock> >::shared_ptr(std::__1::shared_ptr<hpx::components::detail::fixed_wrapper_heap<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, hpx::lcos::local::spinlock> > const&) /opt/apps/x86_64/clang/3.8.0/bin/../include/c++/v1/memory:4292:9
    #1 0x7f2c0b27ec89 in hpx::util::one_size_heap_list<hpx::components::detail::fixed_wrapper_heap<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, hpx::lcos::local::spinlock>, hpx::lcos::local::spinlock>::free(void*, unsigned long) /home/inf3/heller/programming/hpx/hpx/util/one_size_heap_list.hpp:269
    #2 0x7f2c0b2b466e in hpx::components::detail::heap_factory<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> >::free(void*, unsigned long) /home/inf3/heller/programming/hpx/hpx/runtime/components/server/managed_component_base.hpp:322:17
    #3 0x7f2c0b26dad7 in hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>::operator delete(void*, unsigned long) /home/inf3/heller/programming/hpx/hpx/runtime/components/server/managed_component_base.hpp:503:13
    #4 0x7f2c0b2ba099 in std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> >::operator()(hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>*) const /opt/apps/x86_64/clang/3.8.0/bin/../include/c++/v1/memory:2529:13
    #5 0x7f2c0b2ba099 in std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >::reset(hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>*) /opt/apps/x86_64/clang/3.8.0/bin/../include/c++/v1/memory:2735
    #6 0x7f2c0b2ba099 in std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >::~unique_ptr() /opt/apps/x86_64/clang/3.8.0/bin/../include/c++/v1/memory:2703
    #7 0x7f2c0b2ba099 in void hpx::util::detail::invoke_impl<void, hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}>::operator()<{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}, std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > > >({lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}&&, std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&) /home/inf3/heller/programming/hpx/hpx/util/invoke.hpp:35
    #8 0x7f2c0b2ba099 in void hpx::util::invoke_r<void, hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}, std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > > >(hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}&&, std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&) /home/inf3/heller/programming/hpx/hpx/util/invoke.hpp:165
    #9 0x7f2c0b2ba099 in void hpx::util::detail::invoke_fused_impl<void, hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}, hpx::util::tuple<std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > > >, 0ul>(hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}&&, hpx::util::tuple<std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > > >&&, hpx::util::detail::pack_c<unsigned long, 0ul>) /home/inf3/heller/programming/hpx/hpx/util/invoke_fused.hpp:73
    #10 0x7f2c0b2b7c83 in hpx::util::detail::invoke_fused_result<hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}, hpx::util::tuple<std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > > > >::type hpx::util::invoke_fused<hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}, hpx::util::tuple<std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > > > >(hpx::util::detail::invoke_fused_result&&, hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1}&&) /home/inf3/heller/programming/hpx/hpx/util/invoke_fused.hpp:102:16
    #11 0x7f2c0b2b7c83 in hpx::util::detail::deferred<hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1} (std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&)>::operator()() /home/inf3/heller/programming/hpx/hpx/util/deferred_call.hpp:103
    #12 0x7f2c0b2b7c83 in void hpx::util::detail::invoke_impl<void, hpx::util::detail::deferred<hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1} (std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&)> >::operator()<hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1} (&)(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&)>(hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1} (&)(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&)) /home/inf3/heller/programming/hpx/hpx/util/invoke.hpp:36
    #13 0x7f2c0b2b7c83 in void hpx::util::invoke_r<void, hpx::util::detail::deferred<hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1} (std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&)>&>(hpx::util::detail::deferred<hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1} (std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&)>&) /home/inf3/heller/programming/hpx/hpx/util/invoke.hpp:165
    #14 0x7f2c0b2b7c83 in void hpx::util::detail::callable_vtable<void ()>::_invoke<hpx::util::detail::deferred<hpx::lcos::detail::promise_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::lcos::detail::promise_data<std::__1::vector<long, std::__1::allocator<long> > > >::init_shared_state()::{lambda(std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >)#1} (std::__1::unique_ptr<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type>, std::__1::default_delete<hpx::components::managed_component<hpx::lcos::detail::promise_lco<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >, hpx::components::detail::this_type> > >&&)> >(void**) /home/inf3/heller/programming/hpx/hpx/util/detail/vtable/callable_vtable.hpp:74
    #15 0xc01cba in hpx::util::detail::function_base<hpx::util::detail::unique_function_vtable<void ()>, void ()>::operator()() const /home/inf3/heller/programming/hpx/hpx/util/detail/basic_function.hpp:196:20
    #16 0xc01cba in hpx::lcos::detail::compose_cb_impl<hpx::util::unique_function<void (), false>, hpx::util::unique_function<void (), false> >::operator()() /home/inf3/heller/programming/hpx/hpx/lcos/detail/future_data.hpp:237
    #17 0xc00dad in void hpx::util::detail::invoke_impl<void, hpx::lcos::detail::compose_cb_impl<hpx::util::unique_function<void (), false>, hpx::util::unique_function<void (), false> > >::operator()<hpx::lcos::detail::compose_cb_impl<hpx::util::unique_function<void (), false>, hpx::util::unique_function<void (), false> >&>(hpx::lcos::detail::compose_cb_impl<hpx::util::unique_function<void (), false>, hpx::util::unique_function<void (), false> >&) /home/inf3/heller/programming/hpx/hpx/util/invoke.hpp:36:21
    #18 0xc00dad in void hpx::util::invoke_r<void, hpx::lcos::detail::compose_cb_impl<hpx::util::unique_function<void (), false>, hpx::util::unique_function<void (), false> >&>(hpx::lcos::detail::compose_cb_impl<hpx::util::unique_function<void (), false>, hpx::util::unique_function<void (), false> >&) /home/inf3/heller/programming/hpx/hpx/util/invoke.hpp:165
    #19 0xc00dad in void hpx::util::detail::callable_vtable<void ()>::_invoke<hpx::lcos::detail::compose_cb_impl<hpx::util::unique_function<void (), false>, hpx::util::unique_function<void (), false> > >(void**) /home/inf3/heller/programming/hpx/hpx/util/detail/vtable/callable_vtable.hpp:74
    #20 0x7f2c0b25eef1 in hpx::util::detail::function_base<hpx::util::detail::unique_function_vtable<void ()>, void ()>::operator()() const /home/inf3/heller/programming/hpx/hpx/util/detail/basic_function.hpp:196:20
    #21 0x7f2c0b25eef1 in hpx::lcos::detail::future_data<std::__1::vector<long, std::__1::allocator<long> > >::run_on_completed(hpx::util::unique_function<void (), false>&&, std::exception_ptr&) /home/inf3/heller/programming/hpx/hpx/lcos/detail/future_data.hpp:513
    #22 0x7f2c0b25d57a in hpx::lcos::detail::future_data<std::__1::vector<long, std::__1::allocator<long> > >::handle_on_completed(hpx::util::unique_function<void (), false>&&) /home/inf3/heller/programming/hpx/hpx/lcos/detail/future_data.hpp:541:22
    #23 0x7f2c0b26c86a in void hpx::lcos::detail::future_data<std::__1::vector<long, std::__1::allocator<long> > >::set_exception<std::exception_ptr const&>(std::exception_ptr const&, hpx::error_code&) /home/inf3/heller/programming/hpx/hpx/lcos/detail/future_data.hpp:661:17
    #24 0x7f2c0b2a953a in hpx::lcos::detail::promise_lco_base<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >::set_exception(std::exception_ptr const&) /home/inf3/heller/programming/hpx/hpx/lcos/detail/promise_lco.hpp:80:17
    #25 0x7f2c1099beab in hpx::lcos::base_lco::set_exception_nonvirt(std::exception_ptr const&) /home/inf3/heller/programming/hpx/src/lcos/base_lco.cpp:52:9
    #26 0x7f2c086a4c23 in void hpx::actions::basic_action_impl<void (hpx::lcos::base_lco::*)(std::exception_ptr const&), void (hpx::lcos::base_lco::*)(std::exception_ptr const&), &hpx::lcos::base_lco::set_exception_nonvirt, hpx::lcos::base_lco::set_exception_action>::invoke<std::exception_ptr>(unsigned long, int, std::exception_ptr&&) /home/inf3/heller/programming/hpx/hpx/runtime/actions/component_action.hpp:64:20
    #27 0x7f2c086daf87 in hpx::util::unused_type hpx::actions::basic_action<hpx::lcos::base_lco, void (std::exception_ptr const&), hpx::lcos::base_lco::set_exception_action>::invoker::invoke<std::exception_ptr>(std::__1::integral_constant<bool, true>, unsigned long, int, std::exception_ptr&&) const /home/inf3/heller/programming/hpx/hpx/runtime/actions/basic_action.hpp:211:17
    #28 0x7f2c086daf87 in hpx::util::unused_type hpx::actions::basic_action<hpx::lcos::base_lco, void (std::exception_ptr const&), hpx::lcos::base_lco::set_exception_action>::invoker::operator()<std::exception_ptr>(unsigned long, int, std::exception_ptr&&) const /home/inf3/heller/programming/hpx/hpx/runtime/actions/basic_action.hpp:201
    #29 0x7f2c086daf87 in hpx::util::unused_type hpx::actions::basic_action<hpx::lcos::base_lco, void (std::exception_ptr const&), hpx::lcos::base_lco::set_exception_action>::execute_function<std::exception_ptr>(unsigned long, int, std::exception_ptr&&) /home/inf3/heller/programming/hpx/hpx/runtime/actions/basic_action.hpp:346
    #30 0x7f2c086daf87 in void hpx::applier::detail::apply_helper<hpx::lcos::base_lco::set_exception_action, true>::call<std::exception_ptr>(hpx::threads::thread_init_data&&, hpx::naming::id_type const&, unsigned long, int, hpx::threads::thread_priority, std::exception_ptr&&) /home/inf3/heller/programming/hpx/hpx/runtime/applier/apply_helper.hpp:164
    #31 0x7f2c086daf87 in bool hpx::applier::detail::apply_l_p<hpx::lcos::base_lco::set_exception_action, std::exception_ptr>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, std::exception_ptr&&) /home/inf3/heller/programming/hpx/hpx/runtime/applier/apply.hpp:196
    #32 0x7f2c0863d178 in bool hpx::detail::apply_impl<hpx::lcos::base_lco::set_exception_action, std::exception_ptr>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, std::exception_ptr&&) /home/inf3/heller/programming/hpx/hpx/runtime/applier/detail/apply_implementations.hpp:199:28
    #33 0x7f2c086354b1 in hpx::set_lco_error(hpx::naming::id_type const&, hpx::naming::address&&, std::exception_ptr&&, bool) /home/inf3/heller/programming/hpx/src/runtime/trigger_lco.cpp:103:13
    #34 0x7f2c08796b60 in hpx::actions::continuation::trigger_error(std::exception_ptr&&) /home/inf3/heller/programming/hpx/src/runtime/actions/continuation.cpp:91:9
    #35 0x7f2c0b2d0350 in void hpx::actions::detail::trigger_impl<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::util::detail::deferred<hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker (unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&)>&>(std::__1::integral_constant<bool, false>, hpx::actions::typed_continuation<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >&&, hpx::util::detail::deferred<hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker (unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&)>&) /home/inf3/heller/programming/hpx/hpx/runtime/actions/trigger.hpp:40:17
    #36 0x7f2c0b2cf880 in void hpx::actions::trigger<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> >, hpx::util::detail::deferred<hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker (unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&)>&>(hpx::actions::typed_continuation<std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long, std::__1::allocator<long> > >&&, hpx::util::detail::deferred<hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker (unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&)>&) /home/inf3/heller/programming/hpx/hpx/runtime/actions/trigger.hpp:68:9
    #37 0x7f2c0b2cf1fb in hpx::actions::detail::continuation_thread_function<hpx::agas::server::primary_namespace::decrement_credit_action, hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker, unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&>::operator()(hpx::threads::thread_state_ex_enum) /home/inf3/heller/programming/hpx/hpx/runtime/actions/basic_action.hpp:106:17
    #38 0x7f2c0b2cf1fb in std::__1::pair<hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> > hpx::util::detail::invoke_impl<std::__1::pair<hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> >, hpx::actions::detail::continuation_thread_function<hpx::agas::server::primary_namespace::decrement_credit_action, hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker, unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&> >::operator()<hpx::actions::detail::continuation_thread_function<hpx::agas::server::primary_namespace::decrement_credit_action, hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker, unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&>&, hpx::threads::thread_state_ex_enum>(hpx::actions::detail::continuation_thread_function<hpx::agas::server::primary_namespace::decrement_credit_action, hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker, unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&>&, hpx::threads::thread_state_ex_enum&&) /home/inf3/heller/programming/hpx/hpx/util/invoke.hpp:36
    #39 0x7f2c0b2cf1fb in std::__1::pair<hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> > hpx::util::invoke_r<std::__1::pair<hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> >, hpx::actions::detail::continuation_thread_function<hpx::agas::server::primary_namespace::decrement_credit_action, hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker, unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&>&, hpx::threads::thread_state_ex_enum>(hpx::actions::detail::continuation_thread_function<hpx::agas::server::primary_namespace::decrement_credit_action, hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker, unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&>&, hpx::threads::thread_state_ex_enum&&) /home/inf3/heller/programming/hpx/hpx/util/invoke.hpp:165
    #40 0x7f2c0b2cf1fb in std::__1::pair<hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> > hpx::util::detail::callable_vtable<std::__1::pair<hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> > (hpx::threads::thread_state_ex_enum)>::_invoke<hpx::actions::detail::continuation_thread_function<hpx::agas::server::primary_namespace::decrement_credit_action, hpx::actions::basic_action<hpx::agas::server::primary_namespace, std::__1::vector<long, std::__1::allocator<long> > (std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > >), hpx::agas::server::primary_namespace::decrement_credit_action>::invoker, unsigned long&, int&, std::__1::vector<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type>, std::__1::allocator<hpx::util::tuple<long, hpx::naming::gid_type, hpx::naming::gid_type> > > const&&> >(void**, hpx::threads::thread_state_ex_enum&&) /home/inf3/heller/programming/hpx/hpx/util/detail/vtable/callable_vtable.hpp:74
    #41 0x7f2c09acaad2 in hpx::util::detail::function_base<hpx::util::detail::unique_function_vtable<std::__1::pair<hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> > (hpx::threads::thread_state_ex_enum)>, std::__1::pair<hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> > (hpx::threads::thread_state_ex_enum)>::operator()(hpx::threads::thread_state_ex_enum) const /home/inf3/heller/programming/hpx/hpx/util/detail/basic_function.hpp:196:20
    #42 0x7f2c09acaad2 in hpx::threads::coroutines::detail::coroutine_impl::operator()() /home/inf3/heller/programming/hpx/src/runtime/threads/coroutines/detail/coroutine_impl.cpp:96
    #43 0x7f2c08d3892f in void hpx::threads::coroutines::detail::lx::trampoline<hpx::threads::coroutines::detail::coroutine_impl>(hpx::threads::coroutines::detail::coroutine_impl*) /home/inf3/heller/programming/hpx/hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp:102:13
```

**Comment by sithhell** [comment]
After investigating a bit more, the error comes from a race between calling `close` on the channel and pushing a new value. This might lead to an exception being thrown when calling `free_resources.set(target.get());` once the channel has been closed, that is, the continuation is run after `free_resources.close(true)` has been called. To avoid this, you should either wait on the future returned by `result.then(...)` or catch the exception that might get thrown when setting a new value in the channel.
Nevertheless, an uncaught exception should not lead to such an abnormal program termination.

**Comment by sithhell** [comment]
Ok, my initial analysis turned out to be wrong.
The problem was a dangling reference:
```
result.then([&free_resources, target](hpx::shared_future<int>){
-------------^^^
```
The captured reference here might outlive the variable if the continuation is called after `hpx_main` returns. This happens since we don't wait on the continuation to be completed. This can be fixed by either explicitly waiting on it or by capturing free_resources by value:
https://github.com/STEllAR-GROUP/hpx/commit/90dc03dfa1c9a2860c7f2e92959f8493cfc7f740#diff-d4f980e843fee7a784b4f35235ff2829R50

Does this fix your issues?

**Comment by hkaiser** [comment]
@sithhell Nice catch! Thanks!

**Comment by krivenko** [comment]
@sithhell Thank you very much for spending your time to analyze what was essentially my mistake!

> This happens since we don't wait on the continuation to be completed.

I was under a wrong impression, that at the point where `hpx::finalize()` is called the reference is still valid, and `hpx::finalize()` waits for all still running continuations.

Anyhow, my example does not crash anymore. Feel free to close this issue.

**Comment by hkaiser** [comment]
`hpx::finalize()` does not wait. It just signals to the runtime that it is ok to shut down. The runtime will make sure all pending tasks are run to completion and will shut down eventually.
